Improve wrapper features

roigcarlo opened the pull request.

roigcarlo commented:
@nathanfranklin It looks like automatic Kratos releases are compiled with python3, which is making kratos to load incorrectly. I'm currently compiling a version with python 2.7.4, but it may take a while.

nathanfranklin commented on use_cases/CFD_example.gid/test_kratosWrapper.py:
```diff
@@ -39,7 +49,7 @@
 CFDWrapper.add_mesh(mesh)
 
-for i in xrange(0,10):
+for i in xrange(0,CFDWrapper.CM[CUBA.NUMBER_OF_TIME_STEPS]):
```
I think the '''CFDCFDWrapper.CM[CUBA.NUMBER_OF_TIME_STEPS]''' should
be describing the number of time steps that occur when '''run''' occurs.  

For example:

```
CFDWrapper.CM[CUBA.NUMBER_OF_TIME_STEPS] = 5
for i in xrange(0, 10):
  CFDWrapper.run()  # runs 5 steps

#has now run a total of 50 steps
```

roigcarlo commented:
hi @nathanfranklin,

I fixed your comment and kratos is ( finally ) running on travis.

We still have some problems with the DEM tests from our side ( it runs correctly but the results are not the expected due to a difference between the test case version and kratos version ). As soon as it gets solved i will ping you.

nathanfranklin commented on wrappers/kratosWrapper.py:
```diff
@@ -51,68 +47,127 @@ def __init__(self):
+        message = 'KratosWrapper does not handle particle container'
+        raise NotImplementedError(message)
+
+    def add_mesh(self, src):
```
This deviates from the ABCModelingEngine as:
1) it does not return anything
2) just keeps the original 'src' mesh.  

For example, (2) would be a problem:

```
 with closing(H5CUDS.open("foo.cuds", 'r')) as file_handle:   
        wrapper.add_mesh(file_handle.get_mesh("meshA"))
        wrapper.run()
```

something along the lines of this should be done:

```
def add_mesh(self, src):
    copy_mesh = Mesh(name=src.name)
    copy_mesh.data = src.data
    copy_mesh.data = H5Mesh(group, self._handle)
    for point in src.iter_points():
         copy_mesh.add_point(point)

    ...
    ...
    self.meshes[src.name] = copy_mesh
    return copy_mesh
```

nathanfranklin replied:
Note, my suggestion of what to be done is also faulty as it wouldn't handle renaming of the mesh (`mesh.name="new_name"`) as the wrapper would be unaware of such changes:

roigcarlo commented:
@nathanfranklin Thx for the feedback. I see that add_mesh was not behaving as expected. I have changed the way the wrapper handles the meshes so it acts according to the description.
I have also fixed the problems with travis ( again, hope it doesn't crash anymore) so integration should start to appear in green.

I am still changing the iter_mesh functions as I see that it does not support iterating over subsets. 
As soon as I have it I will ping you.

itziakos commented:
@roigcarlo, you might want to use the mesh check templates when testing the Mesh behaviour.
you can see example usage:
- https://github.com/simphony/simphony-common/blob/master/simphony/io/tests/test_h5_mesh.py
- https://github.com/simphony/simphony-mayavi/pull/66/files#diff-93100555da9d1e915df12a581b5f457eR1

roigcarlo commented:
@itziakos I'll take a look thx!

itziakos commented:
for an engine creating a mesh whould probably be:

``` python
def setUp(self):
      self.engine = Engine()
      # .... execute the parent setUp()

def create_container(self, name):
      mesh = Mesh(name=name)
      return self.engine.add_mesh(mesh)
```

roigcarlo commented:
@itziakos I've been through the examples you sent me but I cannot see how this is applicable to my wrapper.

Basically the wrapper uses a standard cuds mesh. When the run function is called  the mesh is converted to Kratos format at once and when it has finished to run is converted back to cuds. ( see: . 

https://github.com/simphony/simphony-kratos/blob/improve-wrapper-features/wrappers/kratos_CFD_wrapper.py#L569-L581 

and

https://github.com/simphony/simphony-kratos/blob/improve-wrapper-features/wrappers/kratos_CFD_wrapper.py#L629-L641

Basically what the wrapper exposes is a simphony mesh.

The only way in which I see these mesh test having sense here is, if this is how its supposed to be, to make an intermediate "kratos_mesh.py" implementing the ABCMesh which interacts directly with the kratos data structures element by element rather than converting the mesh at the execution.

itziakos commented:
> Basically the wrapper uses a standard cuds mesh. When the run function is called the mesh is converted to Kratos format at once and when it has finished to run is converted back to cuds. 

I see, the wrapper architecture is similar to a File-IO wrapper but with direct access instead of IO. sorry I missed that. 

No worries then.

nathanfranklin commented on wrappers/tests/dem/test_kratos_dem.py:
```diff
@@ -0,0 +1,67 @@
+import sys
+import os
+
+sys.path.append(os.path.abspath('wrappers/tests/dem'))
```
This causes problem when trying to run the tests outside of `simphony-kratos`.  I will open up an issue to follow this.

nathanfranklin replied:
Issue added #14

nathanfranklin commented on wrappers/tests/dem/test_kratos_dem.py:
```diff
@@ -0,0 +1,67 @@
+from KratosMultiphysics.IncompressibleFluidApplication import *
+from KratosMultiphysics.FluidDynamicsApplication import *
+from KratosMultiphysics.ExternalSolversApplication import *
+from KratosMultiphysics.MeshingApplication import *
```
are these includes from KratosMultiphysics needed?

roigcarlo replied:
Yes. I can try to delete some of them if it is too problematic.

nathanfranklin commented on wrappers/tests/dem/test_kratos_dem.py:
```diff
@@ -0,0 +1,67 @@
+        for i in xrange(0, 1):
+            wrapper.run()
+
+        wrapper.finalize()
```
ABCModelingEngine does not have a`finalize` method.

would it be possible to add whatever it does in the `run()` method?

roigcarlo replied:
Yes, It can be done

nathanfranklin commented on wrappers/tests/dem/test_kratos_dem.py:
```diff
@@ -0,0 +1,67 @@
+            self.rigid_path
+        )
+
+        wrapper.setMeshData(mesh)
```
ABCModelingEngine does not have a `setMeshData` method.

Would it be possible to rename `setMeshData` to `_setMeshData()`
and then have add_mesh() call that first?

roigcarlo replied:
This also can be done by the wrapper internally

nathanfranklin commented on wrappers/kratos_CFD_wrapper.py:
```diff
@@ -535,46 +561,23 @@ def initialize(self):
+        self.importKratosNodes(
             self.get_mesh("Model"),
-            newFluidMp
+            self.fluid_model_part
```
only a mesh named "Model" is ever used?    It appears that if a user adds a mesh with another name (using `add_mesh`), then that mesh is never used.

roigcarlo replied:
Thx for ponting this out. This is really wrong. The problem is that you only can run a fixed number of meshes for each wrapper but of course it does not need to be called "Model".

Would be correct to implement something in the lines of

```
SP[CUBAExt.FLUID_MESH_TO_RUN] = "foo"
```

nathanfranklin commented on wrappers/kratos_CFD_wrapper.py:
```diff
@@ -400,18 +391,42 @@ def __importKratosConditions(self, src, dst, entitylist=None):
 
     def read_modelpart(self, filename):
-        """ Reads a Kratos formated modelpart NYI
+        """ Reads a Kratos formated modelpart
```
I would have expected that this method reads in a KRATOS file and returns a SimPhoNy mesh (with name "Model") .  is that its intent?

just wondering about lines 435-437 below:

```
        self.fluid_model_part = ModelPart("FluidPart")

        self.addNodalVariablesToModelpart(self.fluid_model_part)
```

nathanfranklin replied:
Currently, the user needs to an `read_modelpart` where an existing Kratos ModelPart file is used to create a SimPhoNy mesh that can then be added to the wrapper.

Is there a description anywhere (or is it being planned) that describes the SimPhonyMesh that this wrapper needs.

Maybe an issue could be opened to add documentation about what the Mesh should be like (what CUBA should be used on cells, points etc..) for this kratos wrapper.

nathanfranklin commented on wrappers/kratos_CFD_wrapper.py:
```diff
@@ -615,37 +618,33 @@ def run(self):
         self.time = self.time + Dt
 
-        for n in xrange(0, substeps):
+        for n in xrange(0, self.CM[CUBA.NUMBER_OF_TIME_STEPS]):
```
I like this here....  but it looks like test_kratosWrapper.py (
here https://github.com/simphony/simphony-kratos/pull/9/files#r32045414 ) has not been updated yet

nathanfranklin commented on wrappers/tests/cfd/test_kratos_cfd.py:
```diff
@@ -0,0 +1,75 @@
+        wrapper.setMeshData(mesh)
+        wrapper.add_mesh(mesh)
+
+        for i in xrange(0, wrapper.CM[CUBA.NUMBER_OF_TIME_STEPS]):
```
same as https://github.com/simphony/simphony-kratos/pull/9/files#r32045414

nathanfranklin commented:
@roigcarlo ,  I added some additional comments.  I don't know how best to proceed.  Maybe, take a look at the comments to see if they make sense first.  Then, possibly, fix the easier ones and for ones that will require more time, maybe it would best that individual issues are made.

A lot of the comments which were made for the DEM wrapper apply to the CFD wrapper, and vice versa.

nathanfranklin commented:
@roigcarlo ,  i would also suggest prioritizing #11

roigcarlo commented:
@nathanfranklin Thx for the comments. I apologize for not answering yesterday but it was holyday here and I was not checking my mails. I will start uploading the changes as soon as they pass the tests.

nathanfranklin commented:
Thanks.  Keep me posted.  To include Kratos in the next version of the simphony-framework (so that it can be reported on in D1.5), this PR needs to be merged and and issue #11 and #15 need to be addressed as soon as possible (preferably by Friday). 

if there are any comments for this PR that are difficult to address in this time-frame, then maybe separate issues should be open for them.

roigcarlo commented:
@nathanfranklin I just updated the README.rst.

I'm about to upload a bunch of changes that will crash travis since I have to move to another machine. this hopefully will help to fix #15 by Friday as requested but the code is still a little bit messy

I will ping you as soon as it works again

nathanfranklin commented on setup.py:
```diff
@@ -22,17 +22,30 @@ def write_version_py(filename=None):
+        ]},
     packages=find_packages(),
-    install_requires=["simphony >= 0.1.1"]
+    install_requires=["simphony >= 0.1.5"]
```
could you move this back to >= 0.1.3

(0.1.3 is the goal for integration and I suspect this code runs with 0.1.3, 0.1.4 and 0.1.5)

itziakos commented on setup.py:
```diff
@@ -22,17 +22,30 @@ def write_version_py(filename=None):
+write_version_py(dem)
 
 setup(
-    name='simkratos',
```
I would suggest that you use a more specific name for the kratos plugin (e.g. simphony_kratos). This name is used for the package folder inside the `site-package` and using `wrappers` is very generic and could conflict with other packages.

itziakos commented on setup.py:
```diff
@@ -22,17 +22,30 @@ def write_version_py(filename=None):
 
 setup(
-    name='simkratos',
+    name='wrappers',
```
I would suggest that you use a more specific name for the kratos plugin (e.g. `simphony_kratos`). This name is used for the package folder inside the `site-package` and using `wrappers` is very generic and could conflict with other packages.

itziakos replied:
Please also note that for consistency the main folder will also have to be renamed (e.g. `wrappers -> simphony_kratos`) and the local imports updated as necessary.

itziakos commented on setup.py:
```diff
@@ -22,17 +22,30 @@ def write_version_py(filename=None):
 
 
-write_version_py()
+cfd = os.path.join(
```
You probably do not need to do this. The `version.py` is necessary for the package and it written during setup so that the version of the package is controlled by the setup.py.

itziakos commented on setup.py:
```diff
@@ -5,7 +5,7 @@
     README_TEXT = readme.read()
 
-VERSION = '0.1.0.dev0'
+VERSION = '0.1.5'
```
this is the version of the kratos simphony plugin correct?

itziakos commented on README.rst:
```diff
@@ -15,11 +15,56 @@ simphony-kratos is hosted on github: https://github.com/simphony/simphony-kratos
+
+For Kratos-CFD::
+
+  from simphony.engine import kratos
```
Please update the documentation to reflect the plugin entry points as they are setup by `setup.py`

itziakos commented on .travis.yml:
```diff
@@ -1,9 +1,19 @@
 script:
-  - flake8 --exclude=CFD_example.gid,3balls.gid --ignore=F401,F403 .
\ No newline at end of file
+  - flake8 --exclude=use_cases,utilities,kratos,ProjectParameters.py,DEM_explicit_solver_var.py --ignore=F403 .
```
you should add the specifics into a configuration file, see http://flake8.readthedocs.org/en/latest/config.html#global

nathanfranklin commented on README.rst:
```diff
@@ -15,11 +15,56 @@ simphony-kratos is hosted on github: https://github.com/simphony/simphony-kratos
 Installation
 ------------
 
+Installation is based on setuptools and requeires Python 2.7::
```
requeires -> requires

nathanfranklin commented on README.rst:
```diff
@@ -15,11 +15,56 @@ simphony-kratos is hosted on github: https://github.com/simphony/simphony-kratos
+- `KratosMultiphysics`_
+
+.. _KratosMultiphysics: https://web.cimne.upc.edu/users/croig/data/kratos-simphony.tgz
+
```
Maybe additonal information could be added about downloadable version of Kratos:

This version is prepared for Ubuntu-12-04-LTS 64-bit.  See http://kratos-wiki.cimne.upc.edu/index.php/Download for additional information on getting Kratos for a specific system.

roigcarlo replied:
Sure. I upload it in the next push

roigcarlo commented:
@nathanfranklin Apparently tests are passing now. I believe I have corrected all the suggestions, but please let me double check it again.

After reading @itziakos suggestion both CFD and DEM are under the same import, which I would like to change in a near future. Anyway, this should be compliant with the API so #15 ( I also changed the test so they use `from simphony.engine import kratos`) and #11 should be fixed at this point.

itziakos commented:
> After reading @itziakos suggestion both CFD and DEM are under the same import, which I would like to change in a near future. Anyway, this should be compliant with the API so #15 ( I also changed the test so they use from simphony.engine import kratos) and #11 should be fixed at this point.

@roigcarlo, I did not comment on the multiple entry points lines of the setup.py. There is nothing wrong with that. My comments were specifically about the name of the package (wrappers -> simphony_kratos or something else) and the fact that the setup.py was doing unnecessary things (the version.py file is only related to the package not independent modules).

I feel that there is a confusion and probably we are using different terms. Each simphony plugin is a single python package that contains multiple subpackages and submodules. The name of the kratos plugin was `wrappers` which was a very generic name. The simphony plugin packages can register simphony engine entry points which they discribe in the setup.py. There is no limitation or problem with registering multiple engines from a single package. In sort the name of the kratos package and the name or number of engine entry points is independent.

roigcarlo commented:
@itziakos Thx for the explanation, its very helpful. It is the first time I touched the entire package-module-plugin thing and I am a little confused, but I believe I see your point now. Basically if i understand it correctly we can have:

```
simkratos            package 
→ kratosCFD          plugin1
→ DEMPack            plugin2
```

I think that, for the initial release, it is ok to leave things like they are. I just wanted ( and still want ) to use the multiple entry points to make clear that, even sharing part of the code they are completely separate things. Also I think that it could be interesting to extend the package with other Kratos solvers since they are already there and should be relativelly easy to implement.

Thx again for taking time to explain me the difference.

P.S: Should we merge this to close #11 #14 and #15?

itziakos commented:
> P.S: Should we merge this to close #11 #14 and #15?

I mainly checked the setup.py and travis.yml and did not spent much to time looking at other parts of this large PR. I think that @nathanfranklin is the most appropriate person to answer this since he have been working on the integration and following the changes.

itziakos commented:
> Also I think that it could be interesting to extend the package with other Kratos solvers since they are already there and should be relativelly easy to implement.

Yes, The fact that `kratos` already has a functioning python layer makes it easier.

nathanfranklin commented:
> >  P.S: Should we merge this to close #11 #14 and #15?
> 
> I mainly checked the setup.py and travis.yml and did not spent much to time looking at other parts of this large PR. I think that @nathanfranklin is the most appropriate person to answer this since he have been working on the integration and following the changes.

Looks good to me.  Thanks.

I will integrate on monday and double check if there are any issues.  If all is well, then you make a new release 0.1.1 of Kratos.

nathanfranklin commented:
@roigcarlo ,  all looks good on integration.  just needed to add the mdpa (have commit for you to review at #16).  

After this #9 is merged, can you immediately make a new release (0.1.1 of simphony-kratos). Thanks.

roigcarlo commented:
@nathanfranklin Perfect. I Merged #16. As soon as it passes I will merge this and make a release

nathanfranklin commented:
@roigcarlo thanks.
